### DISCUSSION TITLE
multi-library workspace: Slint struct is duplicated. Can I avoid std::mem::transmute?

### DISCUSSION THREAD

**Discussion by JakubKoralewski** [slint-ui/slint#5999] [Q&A] [answered]
Here is how my project is laid out (pseudocode -- didn't check compilation):

I have two Slint files, one depends on the other:

window.slint
```slint
import { MyProgramState } from "state.slint"
export component MyWindow inherits Window {
	in property <MyProgramState> state;
	...
}
```

state.slint
```slint
export struct MyProgramState {
    logged-in: bool;
}
export component WorkaroundBecauseDoesntGenerateStructWithoutThis inherits Window {}
```

And two Cargo projects inside the same workspace. `window` depends on `state`:

`window` library:

window/lib.rs
```rust
slint::include_modules!();
use state::MyDefault;

fn main() {
   let ui = MyWindow::new().unwrap();
   let my_state = state::MyProgramState::my_default();
   ui.set_state(unsafe {
		std::mem::transmute::<_, crate::MyProgramState>(my_state)
   });
   ui.run()
}
```

window/build.rs
```rust
fn build() {
	slint_build::compile("path-to-window.slint").unwrap();
}
```

`state` library:

state/lib.rs
```rust
slint::include_modules!();

// Some behavior that I want to use MyProgramState for
trait MyDefault {
	fn my_default() -> Self;
}

impl MyDefault for MyProgramState {
	fn my_default() -> Self {
      todo!()
   }
}
```

state/build.rs
```rust
fn build() {
	slint_build::compile("path-to-state.slint").unwrap();
}
```

As you can see right now I have to use mem::transmute to get around this but ideally I'd like to have the MyProgramState defined in only 1 place. Is there a way to accomplish that while keeping multiple libraries?

For me the most elegant solution as the user would be if I could get Slint to generate the MyProgramState struct into the `state` crate, and then somehow use that MyProgramState also inside `window`, but that would require the generated code to import MyProgramState from the `state` crate.

**Comment by hunger** [comment, marked as answer]
I don't think that is possible right now.

I guess that could be done. We would probably need to mark up one of the `MyProgramState` with some meta-info as to which rust type/module it was generated as elsewhere and then replace the code generation with a `use` statement.

We'd also want to generate code for files exporting structs (and probably enums, etc. as well) without containing a component then.
